### PR TITLE
Introduce a concrete result type for JSValue -> Implementation conversions

### PR DESCRIPTION
**Pull request by weinig**
#### 9a160ff054ee88d4ec12d8410a54e5ad42c9270c
<pre>
Introduce a concrete result type for JSValue -&gt; Implementation conversions
<a href="https://bugs.webkit.org/show_bug.cgi?id=274682">https://bugs.webkit.org/show_bug.cgi?id=274682</a>

Reviewed by Ryosuke Niwa.

Introduces &quot;ConversionResult&quot; a template type that serves as the return value
for all the JSValue -&gt; Implementation conversion functions. It acts much like
ExceptionOr, is even implemented similarly on top of Expected, but instead of
storing either a result or the exception information, it stores the result or
a token indicating an exception exists on the JS stack.

The goal of this type is to allow conversions to non-default initializable
types, crucially, dictionaries and unions with Ref members. This introductory
change does not utilize that, but rather has deferred that to a follow up.

This change is mostly comprised of:

- Updating the code generator to handle the new return type.
- Updating JSDOM* converters to return the new type.
- Update non-generated uses of the convert&lt;...&gt; family of functions to handle
  the new result type.

* Source/WTF/wtf/TrailingArray.h:
(WTF::TrailingArray::TrailingArray):
    - Ensure the destructors don&apos;t run on the uninitialized elements when
      initializing trailing array&apos;s from a generator.

* Source/WebCore/Headers.cmake:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
    - Add new files

* Source/WebCore/bindings/scripts/CodeGeneratorJS.pm:
    - Update code generation to handle new conversion result type.

* Source/WebCore/bindings/IDLTypes.h:
    - Add new trait types ConversionResultType and NullableConversionResultType
      to allow customization of the storage in ConversionResult for that type.

    - Also adds a new IDLOptional type that the generator uses to streamline
      optional arguments and optional dictionary members.

* Source/WebCore/bindings/js/JSDOMConvertResult.h: Added.
    - New result type.

* Source/WebCore/bindings/js/JSCallbackData.h:
    - Callback functions and interfaces now generate a JSDOMCallbackConverterTraits
      instance to allow generic mapping from the JS callback type to the implementation
      callback type.

* Source/WebCore/bindings/js/JSDOMConvertAny.h:
* Source/WebCore/bindings/js/JSDOMConvertBase.h:
* Source/WebCore/bindings/js/JSDOMConvertBoolean.h:
* Source/WebCore/bindings/js/JSDOMConvertBufferSource.h:
* Source/WebCore/bindings/js/JSDOMConvertCallbacks.h:
* Source/WebCore/bindings/js/JSDOMConvertDate.h:
* Source/WebCore/bindings/js/JSDOMConvertDictionary.h:
* Source/WebCore/bindings/js/JSDOMConvertEnumeration.h:
* Source/WebCore/bindings/js/JSDOMConvertEventListener.h:
* Source/WebCore/bindings/js/JSDOMConvertInterface.h:
* Source/WebCore/bindings/js/JSDOMConvertJSON.h:
* Source/WebCore/bindings/js/JSDOMConvertNull.h:
* Source/WebCore/bindings/js/JSDOMConvertNullable.h:
* Source/WebCore/bindings/js/JSDOMConvertNumbers.cpp:
* Source/WebCore/bindings/js/JSDOMConvertNumbers.h:
* Source/WebCore/bindings/js/JSDOMConvertObject.h:
* Source/WebCore/bindings/js/JSDOMConvertOptional.h:
* Source/WebCore/bindings/js/JSDOMConvertPromise.h:
* Source/WebCore/bindings/js/JSDOMConvertRecord.h:
* Source/WebCore/bindings/js/JSDOMConvertScheduledAction.h:
* Source/WebCore/bindings/js/JSDOMConvertSequences.h:
* Source/WebCore/bindings/js/JSDOMConvertSerializedScriptValue.h:
* Source/WebCore/bindings/js/JSDOMConvertXPathNSResolver.h:
* Source/WebCore/bindings/js/JSDOMConvertStrings.cpp:
* Source/WebCore/bindings/js/JSDOMConvertStrings.h:
    - Update to return new ConversionResult type.

    - Where there was a lot of unnecessary repetition, such as typed arrays,
      strings and numbers, intermediary converters were added to help reduce
      the overhead.

* Source/WebCore/bindings/js/JSDOMConvertUnion.h:
    - In addition to updating to the new result type, some simplification was
      done using constexpr to replace &quot;Conditional*&quot; meta-functions.

* Source/WebCore/bindings/js/JSDOMConvertVariadic.h:
    - Updated to use the recently introduced FixedVector::createWithSizeFromGenerator,
      which allows creation with non-default constructible arguments that can fail.

* Source/WebCore/Modules/applepay-ams-ui/ApplePayAMSUIPaymentHandler.cpp:
* Source/WebCore/Modules/applepay/paymentrequest/ApplePayPaymentHandler.cpp:
* Source/WebCore/Modules/mediastream/RTCPeerConnection.cpp:
* Source/WebCore/Modules/mediastream/RTCRtpSFrameTransform.cpp:
* Source/WebCore/Modules/mediastream/RTCRtpScriptTransformer.cpp:
* Source/WebCore/Modules/paymentrequest/PaymentRequest.cpp:
* Source/WebCore/Modules/permissions/Permissions.cpp:
* Source/WebCore/Modules/streams/TransformStream.cpp:
* Source/WebCore/Modules/webaudio/AudioWorkletGlobalScope.cpp:
* Source/WebCore/Modules/webtransport/DatagramSink.cpp:
* Source/WebCore/animation/KeyframeEffect.cpp:
* Source/WebCore/bindings/js/InternalReadableStream.cpp:
* Source/WebCore/bindings/js/JSCustomElementRegistryCustom.cpp:
* Source/WebCore/bindings/js/JSDOMWindowCustom.cpp:
* Source/WebCore/bindings/js/JSDocumentCustom.cpp:
* Source/WebCore/bindings/js/JSElementCustom.cpp:
* Source/WebCore/bindings/js/JSElementInternalsCustom.cpp:
* Source/WebCore/bindings/js/JSEventListener.cpp:
* Source/WebCore/bindings/js/JSExtendableMessageEventCustom.cpp:
* Source/WebCore/bindings/js/JSNodeIteratorCustom.cpp:
* Source/WebCore/bindings/js/JSRTCRtpSFrameTransformCustom.cpp:
* Source/WebCore/bindings/js/JSTreeWalkerCustom.cpp:
* Source/WebCore/bindings/js/JSWebAnimationCustom.cpp:
* Source/WebCore/crypto/SubtleCrypto.cpp:
* Source/WebCore/css/CSSStyleSheetObservableArray.cpp:
* Source/WebCore/html/HTMLCanvasElement.cpp:
* Source/WebCore/html/OffscreenCanvas.cpp:
* Source/WebCore/inspector/InspectorCanvas.cpp:
* Source/WebCore/inspector/InspectorCanvasCallTracer.h:
* Source/WebCore/worklets/PaintWorkletGlobalScope.cpp:
* Source/WebKit/WebProcess/Network/WebTransportSendStreamSink.cpp:
    - Update account for new result types.

Canonical link: <a href="https://commits.webkit.org/279321@main">https://commits.webkit.org/279321@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/c2de37b6c0fce94e24bb86182c2102a37d3e4d21

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/53145 "9 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/32482 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/5632 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/56424 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/3868 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/55450 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/39390 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/3599 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/43089 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/2509 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/55243 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/47/builds/30273 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/45878 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/24219 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/27266 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/3205 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/2027 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/46501 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/49114 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/3363 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/58020 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/52658 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/28286 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/3312 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/50492 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/29506 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/46102 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/49798 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/11593 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/30425 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/64963 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/29260 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/12337 "Found 17 new JSC stress test failures: stress/map-iteration.js.dfg-eager, stress/proxy-set-prototype-of.js.mini-mode, stress/put-non-reified-static-function-or-custom.js.dfg-eager, stress/spread-non-array.js.no-llint, wasm.yaml/wasm/function-tests/add-12.js.wasm-eager, wasm.yaml/wasm/function-tests/br-table-as-return.js.wasm-eager, wasm.yaml/wasm/function-tests/f32-const.js.wasm-eager, wasm.yaml/wasm/function-tests/i32-const.js.wasm-eager, wasm.yaml/wasm/function-tests/invalid-duplicate-export.js.wasm-eager, wasm.yaml/wasm/function-tests/loop-sum.js.wasm-eager ... (failure)") |

### CODE REVIEW

File: Source/WebCore/bindings/js/JSDOMConvertResult.h [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    ConversionResultStorage(ConversionResultException token) : value(makeUnexpected(token)) { }
+    ConversionResultStorage(const Type& value) : value(value) { }
+    ConversionResultStorage(Type&& value) : value(WTFMove(value)) { }
+
+    template<typename U>
+    ConversionResultStorage(ConversionResultStorage<U>&& other)
+        : value([&]() -> Expected<Type, ConversionResultException> {
+            if (other.hasException())
+                return makeUnexpected(ConversionResultException());
+            else
```
Comment (rniwa) [inline Source/WebCore/bindings/js/JSDOMConvertResult.h]: Nit: It seems like we don't need to have else here.

File: Source/WebCore/bindings/js/JSDOMConvertResult.h [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    }
+
+    // Special case conversion from T& to T*
+    template<typename U>
+        requires (std::is_pointer_v<Type> && std::is_lvalue_reference_v<U>)
+    ConversionResultStorage(ConversionResultStorage<U>&& other)
+        : value([&]() -> Expected<Type, ConversionResultException> {
+            if (other.hasException())
+                return makeUnexpected(ConversionResultException());
+            else
```
Comment (rniwa) [inline Source/WebCore/bindings/js/JSDOMConvertResult.h]: Ditto.

File: Source/WebCore/bindings/js/JSDOMConvertResult.h [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+
+    ConversionResultStorage(ConversionResultException token) : value(makeUnexpected(token)) { }
+    ConversionResultStorage(Type& value) : value(std::reference_wrapper<Type> { value }) { }
+
+    template<typename U>
+    ConversionResultStorage(ConversionResultStorage<U>&& other)
+        : value([&]() -> Expected<Type, ConversionResultException> {
+            if (other.hasException())
+                return makeUnexpected(ConversionResultException());
+            else
```
Comment (rniwa) [inline Source/WebCore/bindings/js/JSDOMConvertResult.h]: Ditto.

File: Source/WebCore/testing/TypeConversions.idl [hunk 4]
```diff
@@ -100,7 +100,7 @@
     DOMString stringValue = "";
     [LegacyNullToEmptyString] DOMString treatNullAsEmptyStringValue = "";
     sequence<DOMString> sequenceValue = [];
-    (Node or sequence<DOMString> or TypeConversionsOtherDictionary) unionValue = null;
+    (Node? or sequence<DOMString> or TypeConversionsOtherDictionary) unionValue = null;
```
Comment (rniwa) [inline Source/WebCore/testing/TypeConversions.idl:103]: Why the change?

File: Source/WebCore/bindings/js/JSDOMConvertBoolean.h [hunk 5]
```diff
@@ -27,16 +27,24 @@
...
 
 namespace WebCore {
 
 template<> struct Converter<IDLBoolean> : DefaultConverter<IDLBoolean> {
-
     static constexpr bool conversionHasSideEffects = false;
+    using Result = ConversionResult<IDLBoolean>;
 
-    static bool convert(JSC::JSGlobalObject& lexicalGlobalObject, JSC::JSValue value)
+    static ConversionResult<IDLBoolean> convert(JSC::JSGlobalObject& lexicalGlobalObject, JSC::JSValue value)
```
Comment (rniwa) [inline Source/WebCore/bindings/js/JSDOMConvertBoolean.h]: Maybe we can just use Result here since we're defining using?

File: Source/WebCore/bindings/js/JSDOMConvertResult.h (hunk 1 above)
Comment (weinig) [reply Source/WebCore/bindings/js/JSDOMConvertResult.h]: Fixed.

**Review by rniwa — APPROVED** [review]

File: Source/WebCore/bindings/js/JSDOMConvertResult.h (hunk 3 above)
Comment (weinig) [reply Source/WebCore/bindings/js/JSDOMConvertResult.h]: Fixed.

File: Source/WebCore/bindings/js/JSDOMConvertResult.h (hunk 2 above)
Comment (weinig) [reply Source/WebCore/bindings/js/JSDOMConvertResult.h]: Fixed.

File: Source/WebCore/bindings/js/JSDOMConvertBoolean.h (hunk 5 above)
Comment (weinig) [reply Source/WebCore/bindings/js/JSDOMConvertBoolean.h]: Good catch. That was the intent. Fixed.
